Adding NamespacedRules and handling to GR/GRB

JonCrowther opened the pull request:
## Issue: https://github.com/rancher/rancher/issues/42215
 
## New Feature

Adds `NamespacedRules` to the GlobalRole field. Details of the field can be found in the issue, but in summation: 
- When a global role (GR) is created, a role is created in each namespace with the listed rules
- When a global role binding (GRB) is created, a role binding is created in each namespace that binds to the namespaced roles
- When the role/role binding is modified, it checks if it's owned by a GR/GRB and then enqueues it to be verified

The work is split in the following commits:
1. Add the NamespacedRules field
2. Add GRB support
3. Add GR support
4. Add enqueueing
 
## Testing
1. Create a GlobalRole with the new NamespacedRules field
    - It creates Roles based on the NamespacedRules
2. Create a GlobalRoleBinding with the above GlobalRole
    - It creates RoleBindings based on the NamespacedRules of the GR
3. Delete GR/GRB
    - Removes all Roles/RoleBindings

## Engineering Testing
### Manual Testing
<!-- Describe what manual testing you did (if no testing was done, explain why). -->

### Automated Testing
<!-- Ensure there are unit/integration/validation tests added (if possible); describe what cases they cover and do not cover. -->
* Test types added/modified:
    * Unit
    * Integration (Go Framework) **TODO**
    * Integration (v2prov Framework) **TODO**

Summary: Unit tests for all new code and integration tests (**TODO**)

## QA Testing Considerations
<!-- Highlight areas or (additional) cases that QA should test w.r.t a fresh install as well as the upgrade scenarios -->
 
### Regressions Considerations
<!-- Dedicated section to specifically call out any areas that with higher chance of regressions caused by this change, include estimation of probability of regressions -->
This is a new field, so there shouldn't be any regressions. While the update/create functions for GR/GRBs got modified, the new check only occurs when the field is populated.

Existing / newly added automated tests that provide evidence there are no regressions:
* The existing RBAC integration tests prove that the previous RBAC functionality has not been modified

## Security Considerations
Need a security assessment to ensure this doesn't introduce CVEs

JonCrowther commented on pkg/controllers/management/auth/globalroles/enqueue_test.go:
```diff
@@ -9,13 +9,27 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
+func Test_grbGrIndexer(t *testing.T) {
```
I added this test to get 100% code coverage of `enqueue.go`. I didn't make or modify this function

pjbgf commented on pkg/crds/yaml/generated/management.cattle.io_globalroles.yaml:
```diff
@@ -51,6 +51,59 @@ spec:
+                - verbs
+                type: object
+              type: array
+            description: NamespacedRules are the rules that are active in each namespace
+              of this GlobalRole. Does not support *.
```
Please expand on what "Does not support *" mean in this context, as the documentation picks up the underlying types (as per above) which makes it confusing to the end user. It would be nice if we called out any validation that blocks its use, or in case we don't enforce it, what would be the expected side-effects of using `*`.

MbolotSuse replied:
Agree with what Paulo says here, I would change this to something like the following:
```suggestion
            description: NamespacedRules are the rules that are active in each namespace
              of this GlobalRole. * has no special meaning in the keys - these keys are read 
              as raw strings and must exactly match with one existing namespace. 
```

tomleb replied:
Wouldn't it be less confusing to say the key needs to be a valid namespace name (matching the following regex: `[a-z0-9]([-a-z0-9]*[a-z0-9])?`). In that case no need at all to mention `*` (which on first read confused me as to why that would be important to note).

MbolotSuse replied:
I think that a regex in the comment would be more confusing. The key information to convey here is that there is no wildcard support and each key needs to exactly match (no pattern matching) a namespace - I'm not sure that a regex gets us closer to that.

tomleb replied:
I see how my message above might have been confusing, I was not suggesting we literally put the regex in the description. I was suggesting simply that this fields needs to be a valid namespace name. (Then the user can infer by themselves that there is no support for wildcard).

eg: "NamespacedRules are the rules that are active in each namespace of this GlobalRole. Each key must be in the format of a valid namespace name."

But I hadn't even considered that the user my think `foo-*` would work, so maybe it is a good idea to include that in the description as you suggest.

pjbgf commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +595,118 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+			if !reflect.DeepEqual(roleRef, roleBinding.RoleRef) || roleBinding.Labels[grbOwnerLabel] != globalRoleBinding.Name {
+				newRoleBinding := roleBinding.DeepCopy()
+				newRoleBinding.RoleRef = roleRef
+				if newRoleBinding.Labels != nil {
+					newRoleBinding.Labels[grbOwnerLabel] = globalRoleBinding.Name
+				} else {
+					newRoleBinding.Labels = map[string]string{grbOwnerLabel: globalRoleBinding.Name}
+				}
```
Maybe?

```suggestion
				if newRoleBinding.Labels == nil {
					newRoleBinding.Labels = map[string]string{}
				}
				newRoleBinding.Labels[grbOwnerLabel] = globalRoleBinding.Name
```

pjbgf reviewed (commented):
two small nits:

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go:
```diff
@@ -1146,3 +1104,486 @@ func Test_reconcileClusterPermissions(t *testing.T) {
 
 }
+
+func Test_reconcileNamespacedPermissions(t *testing.T) {
```
I think this needs a test case for failing to list the existing bindings.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +595,118 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+						grbOwnerLabel: globalRoleBinding.Name,
+					},
+				},
+				RoleRef: v1.RoleRef{
```
I think you want the rbac api group in here as well. [Example](https://github.com/rancher/rancher/blob/6708b78e4c3e7621963951ec52bfaf0d493a1d02/pkg/impersonation/impersonation.go#L320).

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +595,118 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+}
+
+// remove any roleBindings that aren't in the namespaces listed in the associated GlobalRole.namespacedRules
+func (grb *globalRoleBindingLifecycle) purgeInvalidRoleBindings(rbs []*v1.RoleBinding, namespacedRules map[string][]v1.PolicyRule) error {
```
Nit: If you could name this in a way that distinguishes it from the other purge method (which is only for the inheritedClusterRole logic), that would be preferable.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +595,118 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+				} else {
+					newRoleBinding.Labels = map[string]string{grbOwnerLabel: globalRoleBinding.Name}
+				}
+				_, err = grb.roleBindings.Update(newRoleBinding)
```
Unfortunately, I don't think you can do an update here. According to [k8s docs](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#clusterrolebinding-example) (and manual testing), the roleRef field on the bindings is immutable, so we will need to create/delete if we find that the binding is corrupt.

Since this is required for at least one case, I would recommend doing the create/update if any corruption is detected on the binding.

JonCrowther replied:
Replaced with a delete/create path instead

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +595,118 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+				Subjects: []v1.Subject{subject},
+			}
+			_, err := grb.roleBindings.Create(newRoleBinding)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
```
I think in the case that this already exists, we need to do an integrity check on the existing role, correct? In this case, if we have a corrupt binding (i.e. referring to an invalid role) in place, we won't correct that for a long time because of this logic.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +214,95 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+				Rules: rules,
+			}
+			_, err = gr.rClient.Create(newRole)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
```
I think we need to validate the integrity of the existing role if it already exists. I would recommend removing the else below and allowing the already exist case to fall through to the below logic.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go:
```diff
@@ -1146,3 +1104,486 @@ func Test_reconcileClusterPermissions(t *testing.T) {
+	}
+}
+
+func Test_purgeInvalidRoleBindings(t *testing.T) {
```
Similar comment here about combining these cases with `reconcile`

MbolotSuse reviewed (changes requested):
Couple of standout issues:
- We need to use `name.SafeConcatName` on the label values
- The purge methods should delete roles/bindings in the same namespace that claim ownership by a GR/GRB but aren't actually owned by that GR/GRB
- We can't update the roleRef on a binding, so you need to delete/create for the binding case instead of update

JonCrowther commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +595,118 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get role %s", roleName))
+			continue
+		}
```
Removed

JonCrowther commented:
Newest commit adds the following small reworks:
- Use of `wrangler.SafeConcatName` for the label
- Moved the purge unit tests into the larger reconcile unit tests
- More unit tests that had been missed
- More comments
- Improved explanation of how `*` is not supported
- Through manual testing, realized I needed `DeleteNamespaced` to delete Roles and RoleBindings, not just `Delete`

Large rework:
- Instead of searching for incorrectly made Role and RoleBindings, I create a list of all the correct items (using UID) while creating/updating all the Role/Bindings. After I select all Role/Bindings with the owner label and delete any that are not in my list of UIDs to keep. That way it should be more comprehensive in deleting Role/Bindings and avoid edge cases

pjbgf commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+			if !reflect.DeepEqual(role.Rules, rules) || role.Labels[grOwnerLabel] != globalRole.Name {
+				newRole := role.DeepCopy()
+				newRole.Rules = rules
+				if newRole.Labels != nil {
+					newRole.Labels[grOwnerLabel] = globalRole.Name
+				} else {
+					newRole.Labels = map[string]string{grOwnerLabel: globalRole.Name}
+				}
```
Maybe?
```suggestion
				if newRole.Labels == nil {
					newRole.Labels = map[string]string{}
				}
				newRole.Labels[grOwnerLabel] = globalRole.Name
```

pjbgf commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+
+	// get all the roles claiming to be owned by this GR and remove any that shouldn't exist
+	r, err := labels.NewRequirement(grOwnerLabel, selection.Equals, []string{globalRole.Name})
+	if err != nil {
+		returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't create label: %s", grOwnerLabel))
+	} else {
+		roles, err := gr.rLister.List("", labels.NewSelector().Add(*r))
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't list roles with label %s : %s", grOwnerLabel, globalRole.Name))
+		} else {
+			err = gr.purgeInvalidNamespacedRoles(roles, roleUIDs)
+			if err != nil {
+				returnError = multierror.Append(returnError, err)
+			}
+		}
+	}
```
Maybe simply flow by removing nested ifs:

```suggestion
	if err != nil {
		return multierror.Append(returnError, errors.Wrapf(err, "couldn't create label: %s", grOwnerLabel))
	}
	
	roles, err := gr.rLister.List("", labels.NewSelector().Add(*r))
	if err != nil {
		return multierror.Append(returnError, errors.Wrapf(err, "couldn't list roles with label %s : %s", grOwnerLabel, globalRole.Name))
	}
	
	err = gr.purgeInvalidNamespacedRoles(roles, roleUIDs)
	if err != nil {
		returnError = multierror.Append(returnError, err)
	}
```

pjbgf commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go:
```diff
@@ -1146,3 +1176,674 @@ func Test_reconcileClusterPermissions(t *testing.T) {
+	}
+}
+
+/* func Test_purgeInvalidRoleBindings(t *testing.T) {
```
Was this left commented by mistake?

JonCrowther replied:
Oops...

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+				continue
+			} else {
+				// In the case that the role already exists, we fall through to check that the rules are correct
+				role = newRole
```
Do we want to set role equal to the new role here? If the role already exists, the version of the role which exists in k8s could have different rules than the one constructed in memory here, and in this case we would be checking the integrity of the one constructed in memory.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+}
+
+// purgeInvalidNamespacedRoles removes any roles that aren't in the slice of UIDS that we created/updated in reconcileNamespacedRoles
+func (gr *globalRoleLifecycle) purgeInvalidNamespacedRoles(roles []*v1.Role, uids []types.UID) error {
```
Nit: You may want to consider using a `map[types.UID]struct{}` instead. It gives you quicker lookup times, especially in cases where users have a large number of valid roles.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+		return multierror.Append(returnError, errors.Wrapf(err, "couldn't create label: %s", grOwnerLabel))
+	}
+
+	roles, err := gr.rLister.List("", labels.NewSelector().Add(*r))
```
This strikes me as a bit inefficient - in the standard case (no corrupt roles) we are going to individually get each item and then re-list them again here. Did you consider any other implementations (like processing the return of a single list call)?

JonCrowther replied:
So giving it some thought, I see 2 options to improve efficiency:

1. I create a map `m` of roles with the matching grOwner label. As I go through the namespaces in NamespacedRules, if I find a role that's in `r`, remove it. After the function, I do `purgeInvalidNamespacedRoles` on the remaining items
2. Using the current method (but with a map of uids instead of a slice), before I call `purgeInvalidNamespacedRoles` I check the length of `roles` and `uids`. If they're the same size, I don't need to call `purgeInvalidNamespacedRoles` because they are guaranteed to be the same list.

In case 1, I maintain the map of roles with the matching labels throughout. In case 2 I only look through it if there are invalid Roles that need to be deleted. So in the case that this is called most frequently on GRs with no invalid roles, case 2 would require fewer operations. In the case where there are invalid roles, case 1 is more efficient.

JonCrowther replied:
I personally think that since the majority of the time this is called, there will be no invalid roles, case 2 is preferable. However the savings either way are minute. We're talking about minimizing map lookups and deletes which are O(1) time so I don't feel too strongly either way.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,115 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+			continue
+		}
+
+		roleName := gr.Name + "-" + ns
```
If we end up updating the role name to use `SafeConcatName`, we will need to do it here as well.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,115 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+		}
+
+		subject := rbac.GetGRBSubject(globalRoleBinding)
+		rbName := globalRoleBinding.Name + "-" + ns
```
Similar comment here about the use of `SafeConcatName` for the name of the binding - I don't think that this is dangerous for this particular type, but better safe than sorry.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,115 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+		rbName := globalRoleBinding.Name + "-" + ns
+		roleBinding, err := grb.roleBindingLister.Get(ns, rbName)
+		if err == nil {
+			if !reflect.DeepEqual(roleRef, roleBinding.RoleRef) || roleBinding.Labels[grbOwnerLabel] != globalRoleBinding.Name {
```
Similar comment here about using safeConcatName here when checking the label value.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,115 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+	}
+
+	// get all the roleBindings claiming to be owned by this GRB and remove any that shouldn't exist
+	r, err := labels.NewRequirement(grbOwnerLabel, selection.Equals, []string{globalRoleBinding.Name})
```
Same comment here about using safe concat name for the label requirements.

MbolotSuse reviewed (changes requested):
Couple of comments on missing uses of safeConcatName, as well as some efficiency improvements.

tomleb commented on pkg/controllers/management/auth/globalroles/enqueue.go:
```diff
@@ -126,3 +129,61 @@ func (g *globalRBACEnqueuer) crtbEnqueueGRB(_, _ string, obj runtime.Object) ([]
+		// this Role isn't owned by a GR, no need to enqueue a GR
+		return nil, nil
+	}
+	_, err := g.grCache.Get(grOwner)
```
Is there any way we can add tests to verify that this works correctly? Perhaps not restricted to the enqueue function but the core GR reconciliation. eg: Including tests with a very long named `Role`.

MbolotSuse replied:
Since these and the other handler which uses this item are sort of separate from one another, this would be something that would be better caught as part of an integration test (i.e. change the role make sure the owner was created, do it again with a longer name).

There's already a lot here, so I'm not eager to call that "in scope" for this PR, but that's my thought on where that would go.

tomleb replied:
Ok.

It feels like our current unit tests don't cover enough (this OP comment kinda points to that) and integration tests are not favored/encouraged for many valid reasons. It would be good to find a solution in between for this (obviously out of scope of this PR).

As for tests that covers this, I'd call that "in scope" for this PR if we had an easy way to add them.. But it looks like we don't, so I won't push more for it. Sorry for derailing the first comment, which still needs fixing :+1:

tomleb commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+	roleUIDs := []types.UID{}
+
+	for ns, rules := range globalRole.NamespacedRules {
+		_, err := gr.nsLister.Get("", ns)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
```
The design doc says:

> Will produce an error if the namespace specified in the key does not exist.

which will probably be enforced by a following webhook PR. What happens if a namespace referenced by a `NamespacedRules` is deleted? Do we prevent the namespace to be deleted until no GRs reference it? Or will we keep erroring when the GR is re-enqueued (say via the sync mechanism every X hours for example).

MbolotSuse replied:
> which will probably be enforced by a following webhook PR.

This is correct.

> What happens if a namespace referenced by a NamespacedRules is deleted?

A status should be added to indicate that the namespace was not found.

> Do we prevent the namespace to be deleted until no GRs reference it?

No - namespaces are core types, we try to avoid validation on those types as much as possible.

> Or will we keep erroring when the GR is re-enqueued (say via the sync mechanism every X hours for example).

We will keep erroring out, but this is valid. If the namespace is re-created, then we want that change to be picked up, and the permissions synced to that location.

In short, most of this (the error message) will be addressed by the status issue.

tomleb replied:
> In short, most of this (the error message) will be addressed by the status issue.

Just to be sure I understand correctly, for now we'll output an error message + re-enqueue, but later we'll remove the error message in favor of a status field?

Here's how I see it could be done and you can let me know if that's how you're also seeing it:
1. Add a watcher for namespaces that enqueues GRs related to it.
2. If namespace doesn't exist, update status + don't retrigger an enqueue.

MbolotSuse replied:
> Just to be sure I understand correctly, for now we'll output an error message + re-enqueue, but later we'll remove the error message in favor of a status field?

For now the error message is fine, yes. 

> 1. Add a watcher for namespaces that enqueues GRs related to it.
> 2. If namespace doesn't exist, update status + don't retrigger an enqueue.

The overall idea makes sense, I think we also need to enqueue the GRBs related to a GR so bindings to the new roles get created. I think that enqueuer [already exists](https://github.com/rancher/rancher/blob/e5d04f65455e083de82e37350c4714d122227501/pkg/controllers/management/auth/globalroles/enqueue.go#L35) so no changes should need to be made though. 

@JonCrowther Can you incoporate these changes into your PR? In short, we need the following behavior now:
- An enqueuer which enqueues a GR if a namespace which is in `namespacedRules` is being changed
- If the namespace is not found, you should log the error, and return a `generic.ErrSkip` from `github.com/rancher/wrangler/pkg/generic` so that you don't re-enqueue.
- Update the comment on the linked enqueuer so that it indicates that we re-enqueue bindings for namespaced permissions too.

JonCrowther replied:
One concern I have here is the use of `generic.ErrSkip`. I'm not familiar with it's behaviour, but since we're using `multierror` to catch what could be several errors, my question  is if we have two errors, one being `generic.ErrSkip` and the other being some random error, does it still get re-enqueued?

JonCrowther replied:
After doing some testing, it looks like if you have `generic.ErrSkip` anywhere in the `multierror` it will skip the re-enqueue. So if there are 2 roles, one gets an error related to Namespace not found, and the second gets an error while updating for example, the reconcile won't be re-enqueued.

I see a few options:

1. The above is expected behaviour and we leave it as is
2. At the end, I can check if the `multierror` has more than one error, and if any aren't `generic.ErrSkip` I can remove the `ErrSkip` errors so it re-enqueues
3. I modify wrangler logic to re-enqueue if there are any errors other than `generic.ErrSkip`

I'm leaning towards 3, but changing the logic of wrangler may have unintended consequences and require more QA than it's worth. Any thoughts?

tomleb replied:
3 seems like a very big change and could potentially break stuff so personally I'd not be in favor. Could we just log the error and not append to the `multierror` at all?

JonCrowther replied:
We could, but then it will continue to re-enqueue the GR and keep getting the namespace error

MbolotSuse replied:
Agree with @tomleb on 3, that's too large a change. As far as I see it, the behavior that you want here is:
- If you have a namespace which isn't found, that shouldn't cause a re-enqueue.
- If you have another type of error (for example, unable to make a binding), that should cause a re-enqueue.

You can achieve this in a few ways, including:
- log/ignore the error if it's a namespace not found
- Don't add the ErrSkip to the multierror since it affects the second requirement
- Only return an errSkip if the rest of the multierror is empty.

Overall, for now I would say just log the not found error and don't bother with errSkip. We can look at it more later, but the complexity might be hurting you here more than is helpful.

tomleb commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,115 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+	return returnError
+}
+
+// purgeInvalidNamespacedRulesRBs removes any roleBindings that aren't in the namespaces listed in the associated GlobalRole.namespacedRules
```
```suggestion
// purgeInvalidNamespacedRBs removes any roleBindings that aren't in the namespaces listed in the associated GlobalRole.namespacedRules
```

Adjust comment to be the same as the function name

tomleb commented on pkg/controllers/management/auth/globalroles/globalrole_handler_test.go:
```diff
@@ -0,0 +1,686 @@
+					return nil, nil
+				}
+				state.nsListerMock.GetFunc = func(_ string, name string) (*corev1.Namespace, error) {
+					if name == "namespace2" {
+						return nil, fmt.Errorf("not found")
+					}
+					return nil, nil
```
It's not clear to me what we're trying to assert here.

The reconciler loops over each namespaces (unordered due to map) and tries to reconcile the rules for those namespaces. If any of them fail, it still keeps going.

This test here only check that the first namespace (in most cases, again, unordered map) is created.
I think it would be best to verify that even though an error occur, other namespaces are also reconciled properly after.

Something like:
- ns1 error
- ns2 success

What do you think?

JonCrowther replied:
You are absolutely correct. That was what I was intending to test but like you said, an unordered map doesn't really guarantee that one goes before the others. I'll still swap them like you suggested

tomleb replied:
I think what you could do instead of just swapping would be to have something like:

```go
state.nsListerMock.GetFunc = func(_ string, name string) (*corev1.Namespace, error) {
					if state.getCount == 0 {
						state.getCount += 1
						return nil, fmt.Errorf("not found")
					}
					return nil, nil
```
so you can be sure that the loop fails for the first item.

Of course now your assertions will depend on which role was successfully created.. What do you think of that?

tomleb commented on pkg/controllers/management/auth/globalroles/globalrole_handler_test.go:
```diff
@@ -0,0 +1,686 @@
+type grTestStateChanges struct {
+	t            *testing.T
+	createdRoles map[string]*rbacv1.Role
+	deletedRoles []string
```
I think it would be safer to have deletedRoles be a map instead of a list. My reasoning is that since the reconciler operates on a map[namespace][]rules, the order of the loop is non deterministic.

So if you were adding tests that deletes multiple roles in different namespaces, you could have one day 
- delete from ns1
- delete from ns2

which would work, but then once in a while you'd end up with
- delete from ns2
- delete from ns1
failing the test (because you'd assume deletedRoles[0] is always from ns1), leading to flaky test.

Note that this isn't the case now because only a single delete is ever tested, so it's more of a "let's future proof this" kinda thing.

tomleb commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,115 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+		roleBinding, err := grb.roleBindingLister.Get(ns, rbName)
+		if err == nil {
+			if !reflect.DeepEqual(roleRef, roleBinding.RoleRef) || roleBinding.Labels[grbOwnerLabel] != globalRoleBinding.Name {
+				// Since roleRef is immutable, we have to delete and recreate the RB
+				err = grb.roleBindings.DeleteNamespaced(roleBinding.Namespace, roleBinding.Name, &metav1.DeleteOptions{})
```
More of a question: I understand that an update is not possible. Has this ever caused trouble (eg: in tests or for usres) where a user loses privileges shortly while a GRB is being reconciled?

For example:
1. Delete `RoleBinding`
2. Connectivity between Rancher & k8s fails momentarily
3. Reconnection
4. Finally create the `RoleBinding`

Maybe it's not too much of a problem in practice in 99% of the cases.

JonCrowther replied:
Michael and I discussed something similar where a user could create an incorrect RoleBinding in the gap between delete and create, and we decided that it's a small enough chance that we are ok with it. Especially since any errors cause the reconcile to be re-queued.

tomleb reviewed (changes requested):
Some comments

JonCrowther commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+				role = newRole
+			}
+		}
+		if role != nil {
+			roleUIDs = append(roleUIDs, role.GetUID())
+
```
imo that's cleaner looking than my existing code. I will update it.

JonCrowther commented:
New commit changes:

- Added functionality for handling deleting/creating namespaces in the NamespacedRules field
  - Added an enqueuer to enqueue a GR when a namespace is modified
  - If namespace doesn't exist, warn the user but don't re-enqueue the reconcile
  - If namespace is terminating, don't attempt to create Roles/RoleBindings
  - Tested this manually (as well as unit tests). If the namespace gets created the Roles/RoleBindings get created too.
- Improved the efficiency of purging invalid Roles/RoleBindings by using maps
- Added and improved comments
- Reduced nested ifs by using `continue`s
- Improved unit tests
- Used `safeConcatName` more consistently throughout and used indexers in the enqueuers to get GR/GRBs based on the new `safeConcatName` versions of the label

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,134 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+		return multierror.Append(returnError, errors.Wrapf(err, "couldn't list roles with label %s : %s", grOwnerLabel, globalRoleName))
+	}
+
+	// After creating/updating all Roles, if the number of RBs with the grOwnerLabel is the same as
+	// as the number of created/updated Roles, we know there are no invalid Roles to purge
+	if len(roleUIDs) != len(roles) {
```
Is this comment true? I'm thinking that there could be a case where a valid role was removed and an invalid role was added - in these cases, you would not purge the invalid Roles.

I don't think that this is too concerning since in theory the add/remove should trigger a re-enqueue, which should create the new role, and then trigger the removal here, but I thought it would at least be worth mentioning.

JonCrowther replied:
In order to cause that to happen, the Role would have to be deleted and an invalid Role be created after the original Role was added to the UID map in the above for loop but before this check happens. While possible, I think it's very unlikely. And as you said, creating an invalid Role with the `authz.management.cattle.io/gr-owner` label set to an existing GR triggers another reconcile via `roleEnqueueGR` and that reconcile would fix it.

Would you like me to mention that edge case in the comment? Ditto for GRB

JonCrowther replied:
I believe in that case, I should remove this `if` and let the purge function run on all Roles found from the `list`.

The alternative I can think of is doing a `list` at the beginning, turning the slice of Roles into a map and then as I go through the NamespacedRules, removing them from the map as I create them. Then at the end I would have a list of Roles that shouldn't be there. This has the same above problem though where new Roles could be created after the `List` call. It also isn't any more efficient since I'd have to loop through the slice of Roles to create a map.

By doing the `List` at the very end, I get the most up to date list of Roles and purge any that are invalid. Let me know what you think, or if you have any other suggestions.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler_test.go:
```diff
@@ -0,0 +1,813 @@
+					return nil, nil
+				}
+				state.nsListerMock.GetFunc = func(_, _ string) (*corev1.Namespace, error) {
+					return nil, nil
```
I think this (and other such usages) are causing a nil pointer in your tests - your code assumes that if there was no error, there was a namespace, and attempts to retrieve the status, which is nil, producing a panic.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,127 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+
+	for ns := range gr.NamespacedRules {
+		namespace, err := grb.nsLister.Get("", ns)
+		if apierrors.IsNotFound(err) {
```
Might also be good to check if the returned namespace was nil

```suggestion
		if apierrors.IsNotFound(err) || namespace == nil {
```

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,134 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+
+	for ns, rules := range globalRole.NamespacedRules {
+		namespace, err := gr.nsLister.Get("", ns)
+		if apierrors.IsNotFound(err) {
```
Might be good to check if the returned namespace was nil
```suggestion
		if apierrors.IsNotFound(err) || namespace == nil{
```

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler_test.go:
```diff
@@ -0,0 +1,813 @@
+			wantError:  true,
+		},
+		{
+			name: "namespace is not found",
```
Would also be good to have a test case which evaluates terminating namespaces.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,127 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+			continue
+		}
+
+		if namespace.Status.Phase == corev1.NamespaceTerminating {
```
Similar comment here about the dangers of a terminating namespace.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -585,6 +598,127 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
+			errors.Wrapf(err, "couldn't list roleBindings with label %s : %s", grbOwnerLabel, grbName))
+	}
+
+	// After creating/updating all RBs, if the number of RBs with the grbOwnerLabel is the same as
+	// as the number of created/updated RBs, we know there are no invalid RBs to purge
+	if len(rbs) != len(roleBindingUIDs) {
```
Same comment here about the dangers of not purging based on the number of active items.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:
```diff
@@ -84,6 +89,7 @@ type globalRoleBindingLifecycle struct {
 	crtbCache         mgmtcontroller.ClusterRoleTemplateBindingCache
 	crtbClient        v3.ClusterRoleTemplateBindingInterface
 	grLister          v3.GlobalRoleLister
+	nsLister          normanv1.NamespaceLister
```
Nit: When adding a new field like this we tend to use the Wrangler caches instead of the Norman ones. If you could swap it I would appreciate it.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -37,6 +49,7 @@ func newGlobalRoleLifecycle(management *config.ManagementContext) *globalRoleLif
 type globalRoleLifecycle struct {
 	crLister rbacv1.ClusterRoleLister
 	crClient rbacv1.ClusterRoleInterface
+	nsLister normanv1.NamespaceLister
```
Nit: When adding a new field like this we tend to use the Wrangler caches instead of the Norman ones. If you could swap it I would appreciate it.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go:
```diff
@@ -1146,3 +1176,554 @@ func Test_reconcileClusterPermissions(t *testing.T) {
+			wantError:         true,
+		},
+		{
+			name: "getting namespace fails",
```
Nit: would be good to have a test case on the terminating namespace.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go:
```diff
@@ -1146,3 +1176,554 @@ func Test_reconcileClusterPermissions(t *testing.T) {
+					return namespacedRulesGR.DeepCopy(), nil
+				}
+				state.nsListerMock.GetFunc = func(_, _ string) (*corev1.Namespace, error) {
+					return nil, nil
```
Similar comment to the GRs on this possibly causing a panic in your tests.

tomleb commented on pkg/controllers/management/auth/globalroles/enqueue.go:
```diff
@@ -126,3 +154,82 @@ func (g *globalRBACEnqueuer) crtbEnqueueGRB(_, _ string, obj runtime.Object) ([]
+
+// roleEnqueueGR enqueues GlobalRoles that own a given Role when that Role is changed. Uses grOwnerLabel
+// which is protected by the webhook rather than the ownerReference.
+func (g *globalRBACEnqueuer) roleEnqueueGR(_, _ string, obj runtime.Object) ([]relatedresource.Key, error) {
```
I'm not familiar enough with wrangler/lasso yet. What happens if the grOwnerLabel is deleted or modified in the `Role` object? Will this enqueuer run for both the old and new object? (old value & new value) Or will nothing be enqueued?

For example, let's say I remove the label and modify the rules by mistake. Do we want Rancher to reconcile the GR again?

JonCrowther replied:
So as far as I understand it, these enqueue methods will run whenever a Role/RoleBinding get modified/created. It then runs the reconciler on any GR/GRBs that get returned. This all happens after the Role/RoleBinding gets modified, so it would be the new version of the Role/RoleBinding.

In your example where the `grOwnerLabel` gets removed, it wouldn't enqueue the GR. However, in a follow up webhook issue, we are adding a check to prevent users from modifying the `grOwnerLabel` https://github.com/rancher/rancher/issues/42216

Modifying the rules goes through the GR/GRB `Updated` functions, which call `reconcileNamespacedRules` so any changes would update the Roles/RoleBindings

tomleb replied:
I see, the webhook will prevent modifying/deleting the owner label.

> Modifying the rules goes through the GR/GRB

I meant modifying the rule of the `Role`, so in that case that would have to go through the `Role`-to-`GlobalRole` enqueuer first. But since  owner label is always valid, no need to worry about this.

Thanks both of you :+1:

tomleb commented on pkg/controllers/management/auth/globalroles/globalrole_handler.go:
```diff
@@ -196,6 +217,134 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+			continue
+		}
+
+		if namespace.Status.Phase == corev1.NamespaceTerminating {
```
From previous testing, `Roles` and `RoleBindings` are deleted (eventually) when a namespace is in terminating state. The namespace doesn't have to be fully removed for the `Roles` / `RoleBindings` to be deleted. So a finalizer **on the namespace** is not enough to keep those objects around (at least, not the way I tested it). For this reason I think it's fine as is, but I don't have a strong opinion on it either.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/enqueue_test.go:
```diff
@@ -497,3 +573,336 @@ func Test_crtbEnqueueGRB(t *testing.T) {
+			wantError: true,
+		},
+		{
+			name: "multiple grs are returned",
```
Is this test case valid? `ns1` was enqueued, but we enqueued two GlobalRoles, neither of which have any rules for `ns1`.

I realize that the indexer, which you are mocking here, is the source of truth for this, but I think that it's a bit odd to have a test case asserting behavior that would be a bug in production.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler_test.go:
```diff
@@ -0,0 +1,834 @@
+		},
+	}
+
+	tests := []struct {
```
Nit: Might be good to have a test case where we have an invalid role in a terminating namespace - we expect no new role to be created, but the existing invalid role should be deleted.

JonCrowther replied:
Added

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrole_handler_test.go:
```diff
@@ -0,0 +1,834 @@
+				}
+				state.nsCacheMock.EXPECT().Get(gomock.Any()).AnyTimes().Return(activeNamespace, nil)
+				state.rListerMock.GetFunc = func(_, _ string) (*rbacv1.Role, error) {
+					if state.counter%2 == 0 {
+						state.counter += 1
+						return nil, apierrors.NewNotFound(schema.GroupResource{
+							Group:    normanv1.RoleGroupVersionKind.Group,
+							Resource: normanv1.RoleGroupVersionResource.Resource,
+						}, "")
+					} else {
+						return nil, fmt.Errorf("error")
+					}
```
Nit: I highly recommend not creating tests cases like this which rely on call order to function. This highly couples your test case to the internal implementation of the underlying code, and can create flaky tests which fail when a specific order occurs.

In this case, I would recommend pre-computing the name of each role that might be retrieved, and then return a specific response for the specific roles which are being called.

JonCrowther replied:
The issue is that the behaviour I'm looking at requires that the `Get` is called twice with the same namespace and role but only fails the second time. I think I could remove the test since it is implementation specific.

JonCrowther replied:
So in this specific test case I could get rid of the counter since I'm trying to get into the error path. However, I kept a similar pattern for the test at line 224. I've added more comments to explain the reason, and changed the counter to be modified based on when `Create` gets called, not just on consecutive `Get` calls.

I considered removing the test at 224, but I think it's important functionality to test. The difficulty is that the `Get` function gets called with the same input twice, but the expected behaviour changes depending on when it happens. I personally think modifying the counter to indicate before and after `Create` makes it less flaky, but if you feel it shouldn't exist I won't argue much.

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go:
```diff
@@ -1146,3 +1175,590 @@ func Test_reconcileClusterPermissions(t *testing.T) {
+			wantError:         false,
+		},
+		{
+			name: "roleBindings get created in terminating namespace",
```
Nit: the name seems a bit innacurate, since we don't want the bindings to be made in a terminating namespace.

JonCrowther replied:
Clarified in the name

MbolotSuse commented on pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go:
```diff
@@ -1146,3 +1175,590 @@ func Test_reconcileClusterPermissions(t *testing.T) {
+		},
+	}
+
+	tests := []struct {
```
Nit: Would be good to have a test case for removing invalid roles from terminating namespaces (and potentially one to ensure that we still create roles in other namespaces should the opportunity arise).

JonCrowther replied:
Added

tomleb commented on pkg/apis/management.cattle.io/v3/authz_types.go:
```diff
@@ -171,6 +171,12 @@ type GlobalRole struct {
+	// * has no special meaning in the keys - these keys are read as raw strings
+	// and must exactly match with one existing namespace.
+	// +optional
+	NamespacedRules map[string][]rbacv1.PolicyRule `json:"namespacedRules,omitempty"`
```
Nit: Should we hint that this field is for the local cluster only or is this _**easily**_ inferred from `InheritedClusterRoles` being "other clusters" and `Rules` being local only?

MbolotSuse replied:
Good point, I think it's worth explicitly stating.

JonCrowther replied:
Added a line to clarify

tomleb reviewed (approved):
No changes since last approval so LGTM. Just added a nit.

JonCrowther commented on pkg/controllers/management/auth/globalroles/enqueue_test.go:
```diff
@@ -9,13 +9,89 @@ import (
+	require.Contains(t, res, "ns1")
+	require.Contains(t, res, "ns2")
+
+	gr.NamespacedRules = map[string][]v1.PolicyRule{}
```
Added

JonCrowther commented on pkg/controllers/management/auth/globalroles/enqueue_test.go:
```diff
@@ -497,3 +573,336 @@ func Test_crtbEnqueueGRB(t *testing.T) {
+			wantError: true,
+		},
+		{
+			name: "multiple grs are returned",
```
I made it more closely reflect the real scenario.

JonCrowther commented:
Rebased and pointed to release/v2.9
